Radicle Link / How it works (#22)
radicle-dev/radicle-docs: docs/understanding-radicle/how-it-works.md

version control backend. It aims to be general enough to be used on top of
systems such as pijul or mercurial, though it's initial implementation is
focused on supporting Git.

The protocol disseminates Git repositories via gossip-based replication,
enabling the hosting and sharing of repositories without reliance on central
servers. Repositories on the Radicle network are called 'projects', which are
gossiped by 'peers'.

In Radicle:

* Peers track other peers
* Peers track projects they are interested in
* Peers gossip about projects. This means replicating updates from the peers
  they track and the projects they are interested in

These interactions create a "trusted" social graph of peers and projects that
become the foundation for collaboration within Radicle.

Radicle Link supports a bazaar-style collaboration model in which there is no
single canonical 'master' branch that contributors merge into, but a multitude
of upstreams exchanging patches via remotes.

<details>
  <summary><b>DISCLAIMER 🌱</b></summary>
  <br>
  <p>
    <em>
      While this documentation reflects the specification of the Radicle Link

>       protocol, note that while all information is implemented, not all

Reviewer: ```suggestion
      protocol, not all
```

rules as per [RFC8141][r8] apply.

An example of a Radicle URN:
`rad:git:hwd1yredksthny1hht3bkhtkxakuzfnjxd8dyk364prfkjxe4xpxsww3try`.

In the Upstream client, Radicle URNs are called [**Radicle IDs**][ri].

Read more [here][ur]

### Delegations

As described in [Data Model][dm] Radicle Link distinguishes two types of
identities: personal and project. The first describes a person (usually denoted
as a `peer`) in the system, while the second describes a software project
(repository) on which one or more persons collaborate. Apart from their
[payload][dp] types `T`, they differ in their delegations type `D`:

Personal identities can only delegate to anonymous keys, while project
identities may attach a personal identity to a key delegation. More formally:

```rust
type User<T> = Identity<T, HashSet<PublicKey>>;

enum ProjectDelegation<U> {
    Key(PublicKey),
    User(User<U>),
}

type Project<T, U> = Identity<T, ProjectDelegation<U>>;

```

The `delegations` of a `Project` are also referred to as the project's
**maintainers**.

These delegations are used to verify the authenticity of data across the
network. See more notes on Verification [here][vr].

### Key Architecture

> 
> With delegations, any identity in Radicle can be verifiable by public-key
> cryptography. Keys on Radicle are architected as such:
> 
> * A `DeviceKey` is an Ed25519 keypair tied to a peer's device (`d`)
> 
> * A `PeerID` is the textual representation of a peer's `DeviceKey`
> 
> In Radicle, keys are securely exchanged with an elliptic curve Diffie-Hellman

Author: ```suggestion
```

> key agreement scheme, where DHKey is used for authentication (as the static
> public key s in the [Noise](http://www.noiseprotocol.org/) protocol).
> 
> The `CodeSigningKey` used in verification may be equal to the `DeviceKey` for

Author: ```suggestion
```

> any device `d` (i.e. as long as the key is not shared between multiple devices).
> Typically, PGP keys already present on a PKI are used for code signing, and
> shared across devices.

Author: ```suggestion
```

are interested in. This happens when a peer clones another peer's project or
tracks a peer directly by adding them as a remote to their project via Upstream.

Since peers represent seperate devices in the network, they each have their own
view of the network. Each peer tracks this view of projects, identities, and
data from connected peers in its own monorepo (See [Git Implementation][gi]).

When a peer tracks another peer in the context of a project — say, if it clones
another peer's project — it sets the intention to fetch and gossip the other
peer's view of that project. This means includes the project metadata, all
working branches and commits, and changesets will be replicated and stored in
the tracking peer's monorepo, so that it can be fetched and collaborated on.

Specifically, this means the tracked peer will be added as a *remote* within the
tracking peer's monorepo as a `$TRACKED_PEER_ID`, as seen in the following
example:

```rust
# Owner of the monorepo
let PEER_ID;

# Peer tracked by $PEER_ID, either directly or transitively
let TRACKED_PEER_ID;

# Identity hash of the project or user
let IDENTITY;

# Identity hashes of certifiers of $IDENTITY
let CERTIFIER[1..];

> $PEER_ID/refs/

Reviewer: Is it intentional that this appears here again?

identities: personal and project. The first describes an person (usually denoted
as a `peer`) in the system, while the second describes a software project
(repository) on which one or more persons collaborate. Apart from their
[payload][dp] types `T`, they differ in their delegations type `D`:

Personal identities can only delegate to anonymous keys, while project
identities may attach a personal identity to a key delegation. More formally:

```rust
type User<T> = Identity<T, HashSet<PublicKey>>;

enum ProjectDelegation<U> {
    Key(PublicKey),
    User(User<U>),
}

type Project<T, U> = Identity<T, ProjectDelegation<U>>;

```

The `delegations` of a `Project` are also referred to as the project's
**maintainers**.

These delegations are used to verify the authenticity of data across the
network. See more notes on Verification [here][vr].

### Key Architecture

With delegations, all identity in Radicle can be verifiable by public-key
cryptography. Keys on Radicle are architected as such:

* A `DeviceKey` is an Ed25519 keypair tied to a peer's device (`d`)

* A `PeerID` is the textual representation of a peer's `DeviceKey`

> In Radicle, keys are securely exchanged with an elliptic curve Diffie-Hellman

Reviewer: This is no longer accurate -- we are using TLS with a self-signed (by the `DeviceKey`) certificate (until someone comes up with an implementation of noise as the QUIC crypto).

key agreement scheme, where DHKey is used for authentication (as the static
public key s in the [Noise](http://www.noiseprotocol.org/) protocol).

> The `CodeSigningKey` used in verification may be equal to the `DeviceKey` for

Reviewer: This paragraph should be removed -- we are not currently devising how code signing is supposed to work, or if it is necessary at all.

Author: @kim - should I just remove the whole Key Architecture section and we can link to your updated spec on this one?

Reviewer: Yeah sounds good

        to * The current and parent attestations refer to different roots

*See more notes on [Verification][ve]*

## Git Implementation

### Overview

Radicle basically uses Git as a database. This means everything is stored in
single Git monorepo that is read and written from via the Upstream client. Our
Git implementation was devised to create an incentive for the seeder to provide
all data necessary to resolve and verify a repository, while reducing latency by
eliminating gossip queries and git fetches as much as possible. 

Read more about our approach [here][ir]

### Namespacing

Radicle uses [namespaces][ns] to store a `peer`'s Radicle data. By using
namespacing, Radicle Link partitions a single git repository - or monorepo -
into logical, smaller repos, that can be checked out individually. This monorepo
includes a `peer`'s identity (personal & project) data, as well as all tracked
Git repositories.

The namespacing scheme has the following format:

```rust
# Owner of the monorepo
let PEER_ID;

# Peer tracked by $PEER_ID, either directly or transitively
let TRACKED_PEER_ID;

# Identity hash of the project or peer
let IDENTITY;

# Identity hashes of certifiers of $IDENTITY

> let CERTIFIER[1..];

Reviewer: `s/CERTIFIER/DELEGATE/`

It is the personal identities a _project_ delegates to (nb. a _personal_ identity can only delegate to _keys_, not other identities).

Reviewer: There is simply no `ids` then, because there is no link to follow.

Reviewer 2: Oh ya good point :+1:

Author: So, what needs to be changed here?

repositories of other peers (this is called a remote in git terminology: a named
reference to a remote repository). If the remote repository is found to track
other remotes, the tracking repository will also transitively track those, up to
n degrees out.

Therefore, a project on Radicle preserves the transitivity information of its
remotes (i.e. via which tracked PeerID another PeerID is tracked).

### Tracking
Tracking is the backbone of collaboration as it drives the exchange of projects
and their artifacts. In Radicle, peers track other peers and projects that they
are interested in. This happens when a peer clones another peer's project or
tracks a peer directly by adding them as a remote to their project via Upstream.

Since peers represent seperate devices in the network, they each have their own
view of the network. Each peer tracks this view of projects, identities, and
data from connected peers in its own monorepo (See [Git Implementation][gi]).

When a peer tracks another peer in the context of a project — say, if it clones
another peer's project — it sets the intention to fetch and gossip the other

> peer's view of that project. This means all includes the project metadata, all

Reviewer 2: ```suggestion
peer's view of that project. This means includes the project metadata, all
```

working branches and commits, and changesets will be replicated and stored in
the tracking peer's monorepo, so that it can be fetched and collaborated on.

Specifically, this means the tracked peer will be added as a *remote* within the
tracking peer's monorepo as a `$TRACKED_PEER_ID`, as seen in the following
example:

```rust
# Owner of the monorepo
let PEER_ID;

# Peer tracked by $PEER_ID, either directly or transitively
let TRACKED_PEER_ID;

# Identity hash of the project or user
let IDENTITY;

# Identity hashes of certifiers of $IDENTITY
let CERTIFIER[1..];

$PEER_ID/refs/
`-- namespaces

>     `-- $PROJECT_NAME

Reviewer 2: ```suggestion
    `-- $IDENTITY
```

alternatively, we use an example where the $ variables are replaced with legitimate values.

leader-based workflow in which they are converging histories of contributing
peers into their main branch. Since their view is verifiable and implicitly
tracked whenever a peer follows a project, peers can ensure they are replicating
a canonical and updated view of a project.

In addition to this, the way Radicle Link works introduces certain implications
for end-user collaboration experience:

- **Your social graph determines what type of content you see, interact with and
  replicate.**

Assuming you have discovered a project of interest within the Radicle network
(more on discoverability later), then the first thing you have to do in order to
interact with it is to track it. Tracking a project signals interest, and by
design implies tracking the project's maintainers, therefore replicating the
data within their social graphs. 

In the context of a project, maintainers of a repository may choose to track the
views of other owners (this is called a remote in Git terminology: a named
reference to a remote repository). If the remote repository is found to track
other remotes, the tracking repository shall also transitively track those, up

> to a configurable N degrees out (currently in the works).

Reviewer 2: ```suggestion
to a configurable N-degrees out (currently in the works).
```

- **Spam and content moderation is naturally handled by the peer's social
  graph**

While this might appear confusing at first, in fact its far more natural (it
actually mimics real life communication) and by design addresses issues like
spam and content moderation, which are naturally handled by the peer's social
graph.

A spammer's patches or issues will never be tracked by the actual maintainers
and as a result they wont be seen by the rest of the network (unless explicitly
tracked). Similarly, if you are not interested in a peer's views or
contributions to a project, you can simply un-follow them, stopping to
replicate, view and interact with their data.

- **Within the same project, two peers might have diverging views.**

The above design also means that even within the same project, peers have
subjective (and often diverging) views.

At minimum, your view of a project becomes the sum of the views of the people
you follow, plus the views of the maintainers of the project. In addition, you
can expand your perspective by configuring your replication settings to also

> transitively track other remotes N degrees out from the peers you follow (ie

Reviewer 2: ```suggestion
transitively track other remotes N-degrees out from the peers you follow (i.e.
```

peers of your peers / remotes of your remotes).

This design also addresses a significant problem with decentralized systems
relying exclusively on distributed ledger technology, the problem of "blockchain
poisoning". This is when someone deliberately adds illegal content to an append
only source in hopes to make the sole act of replicating the project legally
problematic, as correctly pointed out by Konstantin Ryabitsev of the Linux
foundation with regards to a [previous version](http://alpha.radicle.xyz) of
Radicle that was relying on [IPFS](https://ipfs.io/).

Read more about the [Radicle collaboration model][rc]. 

[dl]: #delegations
[dm]: #data-model
[dp]: #doc-payload
[fe]: #fetching
[gi]: #git-implementation
[gs]: getting-started/getting-started.md/#joining-a-seed-node
[ka]: #key-architecture
[na]: #namespacing
[mr]: using-radicle/pushing-and-pulling-changes.md/#adding-remtoes
[ri]: glossary.md/#radicle-id
[rs]: using-radicle/running-a-seed-node.md
[ru]: #radicle-urns
[se]: #serialisation
[vr]: #verification

[sg]: /img/socialgraph.png

[cr]: https://github.com/radicle-dev/radicle-link

> [de]: https://github.com/radicle-dev/radicle-link/blob/master/docs/spec/identities.md#delegations

Reviewer: Best to always use the "permalink" instead of `master` (like [`sp`] below)
